### ISSUE TITLE
Packages depending on ninja to build, cannot be loaded if ninja is uninstalled

### ISSUE THREAD

**Issue by michaelkuhn** [LLNL/spack#37704]
### Steps to reproduce

```console
$ spack install glib
$ spack gc
# This will remove ninja
$ . share/spack/setup-env.sh
$ spack load glib                     
==> Error: spack requires 'ninja'. Make sure it is in your path.
```

### Error message

This seems to be due to ninja's `setup_dependent_package` requiring the ninja binary to be present.

```
==> [2023-05-16-15:38:22.149154] CommandNotFoundError: spack requires 'ninja'. Make sure it is in your path.
==> [2023-05-16-15:38:22.149178] Error: spack requires 'ninja'. Make sure it is in your path.
Traceback (most recent call last):
  File "$SPACK/lib/spack/spack/main.py", line 1018, in main
    return _main(argv)
           ^^^^^^^^^^^
  File "$SPACK/lib/spack/spack/main.py", line 973, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "$SPACK/lib/spack/spack/main.py", line 1001, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "$SPACK/lib/spack/spack/main.py", line 650, in _invoke_command
    return_val = command(parser, args)
                 ^^^^^^^^^^^^^^^^^^^^^
  File "$SPACK/lib/spack/spack/cmd/load.py", line 114, in load
    env_mod.extend(uenv.environment_modifications_for_spec(spec))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "$SPACK/lib/spack/spack/user_environment.py", line 91, in environment_modifications_for_spec
    spack.build_environment.modifications_from_dependencies(
  File "$SPACK/lib/spack/spack/build_environment.py", line 999, in modifications_from_dependencies
    add_modifications_for_dep(dspec)
  File "$SPACK/lib/spack/spack/build_environment.py", line 976, in add_modifications_for_dep
    dpkg.setup_dependent_package(current_module, spec)
  File "$SPACK/var/spack/repos/builtin/packages/ninja/package.py", line 86, in setup_dependent_package
    determine_number_of_jobs(parallel=dspec.package.parallel),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "$SPACK/lib/spack/spack/util/path.py", line 198, in path_filter_caller
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "$SPACK/lib/spack/spack/util/executable.py", line 307, in which_string
    raise CommandNotFoundError("spack requires '%s'. Make sure it is in your path." % args[0])
spack.util.executable.CommandNotFoundError: spack requires 'ninja'. Make sure it is in your path.
```

### Information on your system

* **Spack:** 0.20.0.dev0 (0d991de50af99c9938b3c376cea44cd4656f336e)
* **Python:** 3.11.3
* **Platform:** linux-fedora38-zen2
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by michaelkuhn** [comment]
cc @trws @haampie (who touched this recently)

**Comment by thomas-bouvier** [comment]
I encoutered the same error

* **Spack:** 0.21.0.dev0 (45838cee0b8b01ad999b459ed6d19e37686859c0)
* **Python:** 3.11.4
* **Platform:** linux-fedora38-skylake
* **Concretizer:** clingo

**Comment by eugeneswalker** [comment]
I'm still seeing this issue as of:
```
commit b6f08f1d4ef3b0ca56430c1a465d660828b45ed2 (HEAD -> develop, origin/develop, origin/HEAD)
Author: Vicente Bolea <vicente.bolea@kitware.com>
Date:   Wed Sep 27 12:13:18 2023 -0400

    vtk-m: add VTK-m 2.1.0-rc1 (#39581)
```

Additional info
* **Spack:** 0.21.0.dev0 (b6f08f1d4ef3b0ca56430c1a465d660828b45ed2)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo

**Comment by trws** [comment]
@tgamblin I'll look at it.  This should be a relatively quick fix.

**Comment by tgamblin** [comment]
fixed by #35737

**Comment by BenWibking** [comment]
I am seeing this error, now at build time:
```
==> Installing visit-3.3.3-cqemnbx3hykde6pvf3xopht57chotbbt [113/113]
==> No binary for visit-3.3.3-cqemnbx3hykde6pvf3xopht57chotbbt found: installing from source
==> Error: CommandNotFoundError: spack requires 'ninja'. Make sure it is in your path.

/projects/cvz/bwibking/spack/var/spack/repos/builtin/packages/ninja/package.py:84, in setup_dependent_package:
         81        name = "ninja"
         82
         83        module.ninja = MakeExecutable(
  >>     84            which_string(name, path=[self.spec.prefix.bin], required=True),
         85            determine_number_of_jobs(parallel=dspec.package.parallel),
         86            supports_jobserver=self.spec.version == ver("kitware"),
         87        )
```
